src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3390,44 +3387,10 @@ internal Buffer(IEnumerable<TElement> source, bool queryInterfaces = true)
                 {
                     items = iterator.ToArray();
                     count = items.Length;
-                }
-                else
-                {
-                    ICollection<TElement> collection = source as ICollection<TElement>;
-                    if (collection != null)
-                    {
-                        count = collection.Count;
-                        if (count > 0)
-                        {
-                            items = new TElement[count];
-                            collection.CopyTo(items, 0);
-                        }
-                    }
+                    return;
                 }
             }
-
-            if (items == null)
-            {
-                using (IEnumerator<TElement> e = source.GetEnumerator())
-                {
-                    if (e.MoveNext())
-                    {
-                        items = new TElement[4];
-                        items[0] = e.Current;
-                        count = 1;
-                        
-                        while (e.MoveNext())
-                        {
-                            if (items.Length == count) Array.Resize(ref items, checked(count * 2));
-                            items[count] = e.Current;
-                            ++count;
-                        }
-                    }
-                }
-            }
-
-            this.items = items;
-            this.count = count;
+            items = EnumerableHelpers.ToArray(source, out count);
```

Comment:
#2401 gets rid of it entirely, with iterators not all claiming to be able to produce arrays, and those that can only do so sometimes returning a null that can be checked for when they fail rather than calling back into Buffer.
That was done because separating the ability to create an array (which not all iterators can do anyway) to another interface let the ordered objects implement the same interface, but it might be worth doing anyway.